### PR TITLE
SW-18720 - Add documentation history

### PR DESCRIPTION
**Pull request by janbuecker**
This PR adds a latest changes section to the front page to indicate which articles have been edited recently.

image

### CODE REVIEW

File: app/src/GitHistoryBundle/HistoryListener.php [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        return array(
+            Sculpin::EVENT_BEFORE_RUN => 'dumpGitHistory',
+        );
+    }
+
+    public function dumpGitHistory(SourceSetEvent $event)
+    {
+        /** @var SourceInterface $source */
+        foreach ($event->allSources() as $source) {
+
```
Comment (mitelg) [inline app/src/GitHistoryBundle/HistoryListener.php]: empty line

File: app/src/GitHistoryBundle/HistoryListener.php [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+     */
+    private $permalinkFactory;
+
+    /**
+     * HistoryListener constructor.
+     *
+     * @param string $projectDir
+     * @param $permalinkFactory
+     */
+    public function __construct($projectDir, $permalinkFactory)
```
Comment (PaddyS) [inline app/src/GitHistoryBundle/HistoryListener.php]: Typehint (Sofern wir das hier auch so stark beachten)

File: app/src/GitHistoryBundle/HistoryListener.php (hunk 1 above)
Comment (janbuecker) [inline app/src/GitHistoryBundle/HistoryListener.php]: done

Comment (janbuecker) [inline app/src/GitHistoryBundle/HistoryListener.php]: done

File: app/src/GitHistoryBundle/HistoryListener.php (hunk 2 above)
Comment (janbuecker) [inline app/src/GitHistoryBundle/HistoryListener.php]: done

File: app/src/GitHistoryBundle/HistoryListener.php [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    }
+
+    /**
+     * @param int $numOfCommits
+     *
+     * @return array
+     */
+    private function fetchGitHistory($numOfCommits = 50)
+    {
+        $process = new Process(sprintf("git log --oneline --merges -%d | cut -d' ' -f1", $numOfCommits), $this->projectDir);
```
Comment (bcremer) [inline app/src/GitHistoryBundle/HistoryListener.php]: You might want to `escapeshellarg()` to to escape the shell argument (`$numOfCommits`). It's not necessary yet as the input parameter is in a controlled environment but better future proof it from the beginning.

File: app/src/GitHistoryBundle/HistoryListener.php [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+        if (empty($commits)) {
+            return [];
+        }
+
+        $history = [];
+        $latestHash = 'HEAD^';
+
+        foreach ($commits as $commit) {
+            $getFilesProcess = new Process(sprintf('git diff --name-only %s %s', $commit, $latestHash), $this->projectDir);
```
Comment (bcremer) [inline app/src/GitHistoryBundle/HistoryListener.php]: Possible attack vector here. What happens if I sneak in a commit message containing `;rm /` for example. Please use `escapeshellarg()`.

File: app/src/GitHistoryBundle/HistoryListener.php [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+            $changedFiles = array_filter(
+                explode(PHP_EOL, $getFilesProcess->getOutput()),
+                function ($file) {
+                    if (preg_match('#source/.*\.(md|html)#i', $file)) {
+                        return true;
+                    }
+                }
+            );
+
+            $commitDateProcess = new Process(sprintf('git show --pretty=%%ct %s', $commit), $this->projectDir);
```
Comment (bcremer) [inline app/src/GitHistoryBundle/HistoryListener.php]: `escapeshellarg()`, see above.

File: app/src/GitHistoryBundle/HistoryListener.php [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+     * @param array $sources
+     * @param string $article
+     *
+     * @return null|SourceInterface
+     */
+    private function findArticle(array $sources, $article)
+    {
+        /** @var SourceInterface $source */
+        foreach ($sources as $source) {
+            if (preg_match('#' . $article . '$#i', $source->file()->getPathname())) {
```
Comment (bcremer) [inline app/src/GitHistoryBundle/HistoryListener.php]: `$article` should be escaped to not break the regex: 

```php
preg_match('#' . preg_quote($article, '#') . '$#i'
```

File: app/src/GitHistoryBundle/HistoryListener.php [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    }
+
+    /**
+     * @param SourceSetEvent $event
+     */
+    public function dumpGitHistory(SourceSetEvent $event)
+    {
+        /** @var SourceInterface $source */
+        foreach ($event->allSources() as $source) {
+            if (preg_match('#/source/index.html$#i', $source->file()->getPathname())) {
```
Comment (bcremer) [inline app/src/GitHistoryBundle/HistoryListener.php]: Can be replaced by `if (stripos($source->file()->getPathname(), '/source/index.html' !== false)) {` which is a bit faster.

**Comment by janbuecker** [comment]
Thanks @bcremer! I'll have a look at it after the community day. Hope you're in for a 🍺 !

**Pull request merged**